pep-9999.rst

```diff
@@ -0,0 +1,344 @@
+  * ``PyThread_get_key_value()``
+
+* Misc functions:
+
+  * ``PyBuffer_GetPointer()``: the caller must call
+    ``PyBuffer_Release()``.
+  * ``PyCapsule_Import()``:
+    the caller must hold a reference to the capsule object.
+  * ``Py_GETENV()`` and ``Py_GETENV()`` (``char*``):
+    the pointer can become invalid if environment variables are changed.
+  * ``PyType_GetSlot()``:
+    the caller must hold a reference to the type object.
+  * ``PyModule_GetState()``:
+    the caller must hold a reference to the module object.
+  * ``PyType_GetModuleState()``:
+    the caller must hold a reference to the module object of the type
+    object.
+
+* Deprecated functions, planned for removal:
+
+  * ``Py_GetExecPrefix()``
+  * ``Py_GetPath()``
+  * ``Py_GetPrefix()``
+  * ``Py_GetProgramFullPath()``
+  * ``Py_GetProgramName()``
+  * ``Py_GetPythonHome()``
+
+Backwards Compatibility
+=======================
+
+Only new functions added: existing API are not affected, and no function
+is planned for deprecation.
+
+
+Security Implications
+=====================
+
+Added APIs are safer since they make sure that resource remains valid
+until ``PyResource_Close()`` is being called, and prevent a risk of race
+conditions.
+
+
+How to Teach This
+=================
+
+When a function returns a resource, like a pointer, the
+``PyResource_Close()`` must be called to close the resource.
+
+
+Reference Implementation
+========================
+
+* `Issue #106592 <https://github.com/python/cpython/issues/106592>`_:
+  C API: Add PyResource API: generic API to "close a resource"
+* `PR #107202 <https://github.com/python/cpython/pull/107202>`_
+
+
+Rejected Ideas
+==============
+
+Use the existing PyBuffer_Release() API
+---------------------------------------
+
+The ``Py_buffer`` API already exists and can be modified to fit into
+``PyResource_Close()`` use case. Pseudo-code::
+
+    Py_buffer buffer;
+    char *utf8 = PyUnicode_AsUTF8Res(str, &buffer);
+    // ... use utf8 ...
+    PyBuffer_Release(&buffer);
+
+The ``Py_buffer`` structure has 11 members. They would be initialized
+to:
+
+* ``buf = NULL``
+* ``obj = NULL``
+* ``len = 0``
+* ``itemsize = 1``
+* ``readonly = 1``
+* ``ndim = 1``
+* ``format = NULL``
+* ``shape = NULL``
+* ``strides = NULL``
+* ``suboffsets = NULL``
+* ``internal = NULL``
+
+The problem is that most of these members are irrelevant just to "close
+a resource".  Moreover, the structure has no callback to close the
+resource: it relies on the ``bf_releasebuffer`` protocol of the object
+type (``PyTypeObject.tp_as_buffer.bf_releasebuffer``).
+
+Another problem is that the ``Py_buffer`` structure is now part of the
+stable ABI and so it became complicated to add new members.
+
+Copy the resource at each call
+------------------------------
+
+Instead of caching the UTF-8 encoding string with
+``PyUnicode_AsUTF8()``, the ``PyUnicode_AsUTF8String()`` function can be
+used: it allocates a new string at each call. A similar idea can be
```

Comment:
When would such string be dealloacted? Would it be responsibility of the user?

Reply:
Functions like PyUnicode_AsUTF8String() allocate memory, the caller is responsible to release it.

Reply:
Ah, wow, just read the docs: `PyUnicode_AsUTF8String` returns a new buffer that the user must free, but `PyUnicode_AsUTF8AndSize` returns a cached buffer. Ouch.

Reply:
It makes perfect sense if one thinks about how its implemented, but that's exactly the issue: implementation detail leaks to public API and now it cannot be changed.